### PR TITLE
Fallback to Containerfile when Dockerfile is not present.

### PR DESCRIPTION
**Pull request by manugupt1**
If -f is empty and Dockerfile is not present, look for Containerfile
and build with it.

Implements
https://github.com/containerd/nerdctl/issues/1084#issuecomment-1179466355

Signed-off-by: Manu Gupta <manugupt1@gmail.com>

cc @junnplus  PTAL

### CODE REVIEW

File: cmd/nerdctl/build.go [hunk 1]
```diff
@@ -360,7 +363,16 @@ func generateBuildctlArgs(cmd *cobra.Command, buildkitHost string, platform, arg
 		return "", nil, false, "", nil, nil, err
 	}
 	if _, err := os.Lstat(filepath.Join(absDir, file)); err != nil {
-		return "", nil, false, "", nil, nil, err
+		if file == buildkitutil.DefaultDockerfileName {
+			logrus.Infof("%s not found, checking if Containerfile exists and overriding it", buildkitutil.DefaultDockerfileName)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/build.go:367]: This should be printed only when actually using Containerfile

File: cmd/nerdctl/build.go [hunk 2]
```diff
@@ -360,7 +363,16 @@ func generateBuildctlArgs(cmd *cobra.Command, buildkitHost string, platform, arg
...
+			if _, err := os.Lstat(filepath.Join(absDir, Containerfile)); err == nil {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/build.go:368]: What should we do when `err != nil && !errors.Is(err, os.ErrNotExist)`

File: cmd/nerdctl/build_test.go [hunk 3]
```diff
@@ -42,14 +42,14 @@ CMD ["echo", "nerdctl-build-test-string"]
...
-	base.Cmd("run", "--rm", imageName).AssertOutExactly("nerdctl-build-test-string\n")
-	base.Cmd("run", "--rm", ignoredImageNamed).AssertFail()
+	base.Cmd("--debug", "build", "-t", imageName, buildCtx).AssertOK()
+	// base.Cmd("build", buildCtx, "-t", imageName).AssertOK()
+	// ignoredImageNamed := imageName + "-" + "ignored"
+	// outputOpt := fmt.Sprintf("--output=type=docker,name=%s", ignoredImageNamed)
+	// base.Cmd("build", buildCtx, "-t", imageName, outputOpt).AssertOK()
+
+	// base.Cmd("run", "--rm", imageName).AssertOutExactly("nerdctl-build-test-string\n")
+	// base.Cmd("run", "--rm", ignoredImageNamed).AssertFail()
```
Comment (AkihiroSuda) [inline cmd/nerdctl/build_test.go:52]: ?

File: cmd/nerdctl/build.go (hunk 2 above)
Comment (manugupt1) [reply cmd/nerdctl/build.go]: Returned error in these cases.

File: cmd/nerdctl/build.go (hunk 1 above)
Comment (manugupt1) [reply cmd/nerdctl/build.go]: Added a notice and printed to stdout

File: cmd/nerdctl/build_test.go (hunk 3 above)
Comment (manugupt1) [reply cmd/nerdctl/build_test.go]: Sorry, reverted this change back.

File: cmd/nerdctl/build.go [hunk 4]
```diff
@@ -337,6 +341,7 @@ func generateBuildctlArgs(cmd *cobra.Command, buildkitHost string, platform, arg
 
 	dir := buildContext
 	file := buildkitutil.DefaultDockerfileName
+	const Containerfile string = "Containerfile"
```
Comment (fahedouch) [inline cmd/nerdctl/build.go:344]: May be declaring this const in `buildkitutil`  as we did for defaultDockerfile

File: cmd/nerdctl/build.go [hunk 5]
```diff
@@ -359,9 +364,30 @@ func generateBuildctlArgs(cmd *cobra.Command, buildkitHost string, platform, arg
...
+		if _, err := os.Lstat(filepath.Join(absDir, file)); err != nil {
+			return "", nil, false, "", nil, nil, err
+		}
+	} else {
+		dockerfile, dockerfileerr := os.ReadFile(filepath.Join(absDir, file))
+		containerfile, containerfileerr := os.ReadFile(filepath.Join(absDir, Containerfile))
+		// both files exist.
+		if dockerfileerr == nil {
+			if containerfile == nil && !bytes.Equal(dockerfile, containerfile) {
+				cmd.OutOrStdout().Write([]byte("Dockerfile and Containerfile have different contents, building with Dockerfile"))
```
Comment (fahedouch) [inline cmd/nerdctl/build.go:377]: Why not logrus ?

File: cmd/nerdctl/build.go [hunk 6]
```diff
@@ -359,9 +364,30 @@ func generateBuildctlArgs(cmd *cobra.Command, buildkitHost string, platform, arg
...
+			}
+		} else if dockerfileerr != nil {
+			if errors.Is(err, fs.ErrNotExist) && containerfileerr == nil {
+				logrus.Infof("%s does not exist, using %s as fallback", buildkitutil.DefaultDockerfileName, Containerfile)
+				file = Containerfile
+			} else {
+				return "", nil, false, "", nil, nil, err
+			}
+		} else if dockerfile != nil || containerfile != nil {
```
Comment (fahedouch) [inline cmd/nerdctl/build.go:386]: We will never reach this else if

File: cmd/nerdctl/build_test.go [hunk 7]
```diff
@@ -287,3 +287,27 @@ CMD ["echo", "nerdctl-build-test-string"]
...
+
+	var err = os.WriteFile(filepath.Join(tmpDir, "Containerfile"), []byte(dockerfile), 0644)
+	assert.NilError(t, err)
+
+	buildCtx, err := createBuildContext(dockerfile)
+	assert.NilError(t, err)
+	defer os.RemoveAll(buildCtx)
+
+	base.Cmd("build", "-t", imageName, buildCtx).AssertOK()
+	base.Cmd("run", "--rm", imageName).AssertOutExactly("nerdctl-build-test-string\n")
```
Comment (fahedouch) [inline cmd/nerdctl/build_test.go:312]: Plz add test with both `Dockerfile` and `Containerfile`

File: cmd/nerdctl/build.go (hunk 5 above)
Comment (manugupt1) [reply cmd/nerdctl/build.go]: I added to logrus. @AkihiroSuda would you prefer having it printed out to stdout as well?

File: cmd/nerdctl/build.go (hunk 6 above)
Comment (manugupt1) [reply cmd/nerdctl/build.go]: Fixed this and added unit tests by moving to pkg as there is a lot of logic here.

File: cmd/nerdctl/build.go (hunk 5 above)
Comment (AkihiroSuda) [reply cmd/nerdctl/build.go]: logrus should be enough?

Comment (manugupt1) [reply cmd/nerdctl/build.go]: Okay! Thanks! This change only shows with logrus

File: cmd/nerdctl/build_test.go [hunk 8]
```diff
@@ -287,3 +287,58 @@ CMD ["echo", "nerdctl-build-test-string"]
 	base.Cmd("run", "--rm", imgWithNoTag).AssertOutExactly("nerdctl-build-test-string\n")
 	base.Cmd("run", "--rm", imgWithCustomTag).AssertOutExactly("nerdctl-build-test-string\n")
 }
+
+func TestBuildWithContainerfile(t *testing.T) {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/build_test.go:291]: Why is test passing `[test-integration-docker-compatibility](https://github.com/containerd/nerdctl/runs/7312918381?check_suite_focus=true#logs)` ?
I don't think Docker supports Containerfile

File: pkg/buildkitutil/buildkitutil.go [hunk 9]
```diff
@@ -181,3 +185,42 @@ func WriteTempDockerfile(rc io.Reader) (dockerfileDir string, err error) {
...
+	if err != nil {
+		return "", "", err
+	}
+	if file != DefaultDockerfileName {
+		if _, err := os.Lstat(filepath.Join(absDir, file)); err != nil {
+			return "", "", err
+		}
+	} else {
+		dockerfile, dockerfileerr := ioutil.ReadFile(filepath.Join(absDir, file))
+		containerfile, containerfileerr := ioutil.ReadFile(filepath.Join(absDir, ContainerfileName))
```
Comment (junnplus) [inline pkg/buildkitutil/buildkitutil.go:207]: Only need to read the file content if both files exist. Using `os.Lstat` replace it.

File: pkg/buildkitutil/buildkitutil.go [hunk 10]
```diff
@@ -24,10 +24,13 @@
 package buildkitutil
 
 import (
+	"bytes"
 	"encoding/json"
 	"errors"
 	"fmt"
 	"io"
+	"io/fs"
+	"io/ioutil"
```
Comment (fahedouch) [inline pkg/buildkitutil/buildkitutil.go:33]: Deprecated plz use `os`

File: pkg/buildkitutil/buildkitutil.go [hunk 11]
```diff
@@ -181,3 +185,42 @@ func WriteTempDockerfile(rc io.Reader) (dockerfileDir string, err error) {
...
+			}
+		} else if dockerfileerr != nil {
+			if errors.Is(dockerfileerr, fs.ErrNotExist) && containerfileerr == nil {
+				logrus.Infof("%s does not exist, using %s as fallback", DefaultDockerfileName, ContainerfileName)
+				file = ContainerfileName
+				return absDir, file, nil
+			} else {
+				return "", "", err
+			}
+		} else if containerfileerr != nil {
```
Comment (fahedouch) [inline pkg/buildkitutil/buildkitutil.go:221]: As I said before this else if is never reached

File: pkg/buildkitutil/buildkitutil_test.go [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+   Portions from https://github.com/docker/cli/blob/v20.10.9/cli/command/image/build/context.go
+   Copyright (C) Docker authors.
+   Licensed under the Apache License, Version 2.0
+   NOTICE: https://github.com/docker/cli/blob/v20.10.9/NOTICE
+*/
+
+package buildkitutil
+
+import (
+	"io/ioutil"
```
Comment (fahedouch) [inline pkg/buildkitutil/buildkitutil_test.go:27]: Same as above Depcrecated

File: cmd/nerdctl/build_test.go (hunk 8 above)
Comment (manugupt1) [reply cmd/nerdctl/build_test.go:291]: Fixed; the test also created a docker file.

**Comment by AkihiroSuda** [comment]
Is this still WIP?

**Comment by manugupt1** [comment]
> Is this still WIP?

Fixing a few tests that are failing locally and then it is ready to review.

**Review by AkihiroSuda — APPROVED** [review]
Thanks

File: pkg/buildkitutil/buildkitutil.go [hunk 13]
```diff
@@ -181,3 +184,51 @@ func WriteTempDockerfile(rc io.Reader) (dockerfileDir string, err error) {
...
+			if err != nil {
+				return "", "", err
+			}
+			if !bytes.Equal(dockerfile, containerfile) {
+				logrus.Warnf("%s and %s have different contents, building with %s", DefaultDockerfileName, ContainerfileName, DefaultDockerfileName)
+			}
+		}
+		if dErr != nil {
+			if errors.Is(dErr, fs.ErrNotExist) {
+				logrus.Warnf("%s, using %s as fallback", dErr, ContainerfileName)
```
Comment (Kitanotori) [inline pkg/buildkitutil/buildkitutil.go:223]: @manugupt1 Why does this raise warning? I don't think using Containerfile when Dockerfile doesn't exist should output warning.

**Pull request merged**